nncf/experimental/post_training/api/engine.py

```diff
@@ -19,30 +19,76 @@
 
 ModelType = TypeVar('ModelType')
 TensorType = TypeVar('TensorType')
-
+MetricType = TypeVar('MetricType')
 
 class Engine(ABC):
     """
     The basic class aims to provide the interface to infer the model.
     """
+    # TODO (Nikita Malinin): Update class with the _get_sampler() method
 
     def __init__(self):
         self.model = None
+        self._sampler = None
+        self._data_loader = None
+        self._metrics = None
+
+    # TODO (Nikita Malinin): Add statistic aggregator object (per-backend)
+    
+    @property
+    def data_loader(self):
+        return self._data_loader
+
+    @data_loader.setter
+    def data_loader(self, data_loader):
+        self._data_loader = data_loader
+
+    @property
+    def metrics(self):
+        return self._metrics
+
+    @metrics.setter
+    def metrics(self, metrics):
+        self._metrics = metrics
+
+    @property
+    def sampler(self):
+        return self._sampler
+
+    @sampler.setter
+    def sampler(self, sampler):
+        self._sampler = sampler
 
     def set_model(self, model: ModelType) -> None:
         self.model = model
 
     def is_model_set(self) -> bool:
         return self.model is not None
 
-    def infer(self, _input: TensorType) -> Dict[str, TensorType]:
-        if not self.is_model_set():
-            raise RuntimeError('The {} tried to infer the model, while the model was not set.'.format(self.__class__))
-        return self._infer(_input)
+    def transform_output(self, outputs: TensorType) -> TensorType:
+        """ Processes model output data
+            :param outputs: list of output data
+            :return outputs: list of the output data in an order expected by the accuracy metric if any is used
+        """
+        return outputs
 
-    @abstractmethod
-    def _infer(self, _input: TensorType) -> Dict[str, TensorType]:
+    def transform_input(self, inputs: TensorType) -> TensorType:
+        """ Processes model input data based on the backend
+            :param inputs: list of input data
+            :return inputs: list of the input data
         """
-        Infer the model on the provided input.
-        Returns the model outputs and corresponding node names in the model.
+        return inputs
+
+    def compute_statistics(self, statistics_layout: Dict) -> Dict[str, TensorType]:
```

Comment:
The main point of the methods division is in the UX: in the general pipeline, users shouldn't worry about the implementation of the Samplers or data-iteration loops to collect statistics or calculate metrics based on the raw model output data. 
This is why we provide two different methods (or helpers): 

1. `compute_statistics`, more internal, for statistics calculation because the tool knows which statistics are needed for the current model and with the current quantization settings, which passed should be called before calculation and etc.; 
2. `compute_metrics`, the second helper, for metrics calculation that (will be updated soon) integrates metrics object usage that can be simply re-implemented based on the user's pipeline.

cc @AlexKoff88

Reply:
But in the current design, the user also doesn't worry about Samplers or data-iteration loops to collect statistics or calculate metrics based on the raw model output data. The algorithm does it for them.

Reply:
I would add that Sampler is even not a public class.

Reply:
> I would add that `Sampler` is even not a public class. I mean it is not included in public API

It became public in the current PR. See - nncf/experimental/post_training/api/sampler.py

Reply:
> But in the current design, the user also doesn't worry about Samplers or data-iteration loops to collect statistics or calculate metrics based on the raw model output data. The algorithm does it for them.

And the current implementation still provides the same UX that before. In which scenarios user can use the raw output model data?

Reply:
I see. I wonder why do we make it public? In what cases `RandomBatchSampler` or `BatchSampler` are not enough?

Reply:
> > But in the current design, the user also doesn't worry about Samplers or data-iteration loops to collect statistics or calculate metrics based on the raw model output data. The algorithm does it for them.
> 
> And the current implementation still provides the same UX that before. In which scenarios user can use the raw output model data?

To calculate their own metrcis for example

Reply:
> To calculate their own metrics for example

The current implementation will provide the method `calculate_metrics` that will cal helpers from the `Metic` class like the `update`. This method will get raw outputs and targets for metric calculation.

Reply:
> > To calculate their own metrics for example
> 
> The current implementation will provide the method `calculate_metrics` that will cal helpers from the `Metic` class like the `update`. This method will get raw outputs and targets for metric calculation.

I thing that actually the question is somehow incorrect.
The user basically doesn't work with `Engine`.
The user can only implement own Engine for algorithms usage, right?
And the problem that I see that to implement own Engine in new design getting harder as I should add realisation of more methods, which is not intuitive. And I fell in real life there is no user who will do it.

Reply:
In the general case of re-implementation of their own `Engine`, users update only methods for the data pre- and post-processing.

Reply:
I see. I think that I don't see the main point of the PR:
1) What is the reason for updating the current API? I mean what problems/disadvantages do we see and want to improve.
2) How does this PR solve these problems?